kroxylicious-filters/kroxylicious-sasl-inspection/src/main/java/io/kroxylicious/filters/sasl/inspection/OauthBearerSaslObserver.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filters.sasl.inspection;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import javax.security.sasl.SaslException;
+
+import org.apache.kafka.common.security.oauthbearer.internals.OAuthBearerClientInitialResponse;
+import org.jose4j.jwt.GeneralJwtException;
+import org.jose4j.jwt.consumer.InvalidJwtException;
+import org.jose4j.jwt.consumer.JwtConsumerBuilder;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * A SASL observer capable of extracting an authorization id from a
+ * <a href="https://datatracker.ietf.org/doc/html/rfc7628">OAUTHBEARER</a> initial client response.
+ * <br/>
+ * The authorization id is obtained in the following manner.
+ * If the authorization id is present in the <a href="https://www.rfc-editor.org/rfc/rfc5801.html#page-8">gs2-header</a> it is used.
+ * If there is no authorization id in the header, the subject claim in the JWT payload is used.  If the subject
+ * claim is absent, the payload is encrypted (JWE), or opaque tokens are in use, the authentication will fail.
+ */
+class OauthBearerSaslObserver implements SaslObserver {
+
+    private boolean gotExpectedServerFinal = false;
+
+    private boolean gotServerError = false;
+    @Nullable
+    private String authorizationId = null;
+
+    @Override
+    public String mechanismName() {
+        return "OAUTHBEARER";
+    }
+
+    @Override
+    public boolean clientResponse(byte[] response) throws SaslException {
+        if (authorizationId == null) {
+            var initialResponse = new OAuthBearerClientInitialResponse(response);
+            var authzid = Optional.of(initialResponse.authorizationId()).filter(Predicate.not(String::isEmpty)).map(SaslUtils::username);
+
+            try {
+                if (authzid.isEmpty()) {
+                    // It is not the role of the observer to validate the JWT - the server
+                    // will decide if it is acceptable. We merely need to extract
+                    // the subject claim.
+                    var consumer = new JwtConsumerBuilder()
+                            .setSkipAllValidators()
+                            .setDisableRequireSignature()
+                            .setSkipSignatureVerification()
+                            .build();
+                    var context = consumer.process(initialResponse.tokenValue());
+                    var sub = context.getJwtClaims().getStringClaimValue("sub");
+
+                    authorizationId = Optional.ofNullable(sub).filter(Predicate.not(String::isEmpty))
+                            .orElseThrow(() -> new SaslException("JWT does not contain subject claim (sub)"));
```

Comment:
Does this error message propogate back to the client in a SaslAuthenticate response? If so, should it? It seems to me that this reveals to the client something about how authn is configured (the absence of `sub`). Better to just fail with the generic message.